flang/include/flang/Lower/Mangler.h

```diff
@@ -58,6 +59,38 @@ std::string mangleName(const semantics::DerivedTypeSpec &);
 /// Recover the bare name of the original symbol from an internal name.
 std::string demangleName(llvm::StringRef name);
 
+std::string
+mangleArrayLitteral(const uint8_t *addr, size_t size,
+                    const Fortran::evaluate::ConstantSubscripts &shape,
+                    Fortran::common::TypeCategory cat, int kind = 0,
+                    Fortran::common::ConstantSubscript charLen = -1);
+
+template <Fortran::common::TypeCategory TC, int KIND>
+std::string mangleArrayLitteral(
+    const Fortran::evaluate::Constant<Fortran::evaluate::Type<TC, KIND>> &x) {
+  return mangleArrayLitteral(
+      reinterpret_cast<const uint8_t *>(x.values().data()),
```

Comment:
Is `uint8_t` needed? I'm wondering if these casts could be discarded since `std::string` works with `char*`. (`char` is guaranteed to be exactly 8 bits by the standard, so ...)

Reply:
Val algorithm works by hashing the `Constant<T>` data, which is basically the data of a vector<c++ type for T>. So the data won't work with char* for all T types. `llvm::MD5::update` expects an `llvm::ArrayRef<uint8_t>`, and `char` signedess is platform dependent, so I think using a reinterpret_cast to `uint8_t` is unavoidable with the current algorithm.